[v14.x backport] util: fix module inspection & instanceof check during inspect

Pezmc opened the pull request:
<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

Backports https://github.com/nodejs/node/pull/36178 from @BridgeAR, tagged as semver‑patch, to 14.x as per [comment](https://github.com/nodejs/node/pull/36178#issuecomment-730525754) from @ExE-Boss

14.x is currently impacted by the bug fixed in #36178, which was originally released to 15.x in https://github.com/nodejs/node/releases/tag/v15.5.0 

I followed the guide from https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md

Pezmc commented:
I don't have the Jekins permissions to schedule a node-test-pull-request CI job for this PR (37100)

> Pezmc is missing the Job/Build permission

BethGriggs commented:
Would appreciate a review from someone more familiar with this area (cc @nodejs/modules?)

ljharb commented on lib/internal/util/inspect.js:
```diff
@@ -535,6 +535,14 @@ function getEmptyFormatArray() {
   return [];
 }
 
+function isInstanceof(object, proto) {
+  try {
+    return object instanceof proto;
+  } catch {
+    return false;
+  }
+}
+
```
why is the try/catch needed? are we worried about someone defining a `Symbol.hasInstance` method that can throw?

Pezmc replied:
Based on https://github.com/nodejs/node/pull/36178/commits/cb9eec9f079949c99ac5f32bcd36b90dec07bd11#r527685177, `instanceof` can throw

ljharb replied:
ah, didn't realize this was a backport PR. turns out `instanceof` can indeed throw when the RHS is not a function, or when it's a non-constructible function (like an arrow function or a concise method)

ljharb reviewed (commented):
LGTM; i'll make similar changes in `object-inspect`.

Pezmc commented:
@BethGriggs I [tried rebasing](https://github.com/nodejs/node/compare/v14.x-staging...Pezmc:backport-36178-to-v14.x-test?expand=1) onto the latest `nodejs:v14.x-staging` to clean this PR and ended up with 
89402af in my diff. Do you know where that came from?

```bash
$ git fetch upstream v14.x-staging
From https://github.com/nodejs/node
 * branch                  v14.x-staging -> FETCH_HEAD
$ git checkout --track upstream/v14.x-staging
Updating files: 100% (14760/14760), done.
Branch 'v14.x-staging' set up to track remote branch 'v14.x-staging' from 'upstream'.
Switched to a new branch 'v14.x-staging'
$ git pull
Already up to date.
$ git checkout backport-36178-to-v14.x
Switched to branch 'backport-36178-to-v14.x'
$ git rebase v14.x-staging
Successfully rebased and updated refs/heads/backport-36178-to-v14.x.
$ git checkout -B backport-36178-to-v14.x-test
Switched to a new branch 'backport-36178-to-v14.x-test'
$ git push
```

BethGriggs commented:
> @BethGriggs I tried rebasing onto the latest nodejs:v14.x-staging to clean this PR after your force push and ended up with 89402af in my diff. Do you know where that came from?

Ah yes, that commit was backed out of staging as there was an npm version mismatch (ref: https://github.com/nodejs/node/pull/37107#issuecomment-769260711). We need to reland the npm update in a new PR.

For this PR, I think you could either rebase against `upstream/v14.x-staging` and then drop the npm commit. Or I can just cherry-pick your commits from this PR.

The good news is that the delay in fixing means we can probably squeeze this into v14.15.5 (https://github.com/nodejs/node/pull/37074)

Pezmc commented:
Re-rebased and manually dropped that commit, this PR should be good to go @BethGriggs!

ljharb commented:
@Pezmc did you accidentally force push the branch to nothing? you may need to try to restore the branch, otherwise this PR might need to be recreated.

Pezmc commented:
Accidental close, please ignore!

Pezmc commented:
@BethGriggs Anything I can do to help get this into v14.15.5?

BethGriggs commented:
@pezmc, I'll aim to integrate this into the v14.15.5 proposal - i'm still waiting on some other patches so haven't started updating the proposal yet. Nothing more to do on your side 🙂

BethGriggs commented:
Would appreciate a review on this before landing (maybe @nodejs/modules)?

BethGriggs reviewed (approved):
Backport LGTM comparing with #36178

BethGriggs commented:
Landed in 408c7a65f3...8786f1f448 (v14.x-staging). Thanks @Pezmc

Pezmc commented:
Landed in https://github.com/nodejs/node/releases/tag/v14.15.5! 💃
